lib/spack/spack/cmd/diff.py

```diff
@@ -174,24 +173,58 @@ def group_by_type(diffset):
             cprint("@G{+  %s}" % addition.pop(0))
 
 
-def diff(parser, args):
-    env = ev.active_environment()
+def print_header(a, b):
+    cprint("@R{--- %s}" % a)
+    cprint("@G{+++ %s}" % b)
 
-    if len(args.specs) != 2:
-        tty.die("You must provide two specs to diff.")
 
-    specs = [spack.cmd.disambiguate_spec(spec, env, first=args.load_first)
-             for spec in spack.cmd.parse_specs(args.specs)]
+def _is_environment_comparison(args):
+    is_lock = lambda p: os.path.isfile(p) and '.lock' in os.path.basename(p)
+    return len(args) == 2 and is_lock(args[0]) and is_lock(args[1])
 
-    # Calculate the comparison (c)
+
+def diff(parser, args):
     color = False if args.dump_json else get_color_when()
+
+    # Try to read environment lock files
+    if _is_environment_comparison(args.args):
+        mode = 'env'
```

Comment:
A couple suggestions to clean up the code, the `mode` variable was making it hard to read. It sounded like it was going to be an argument to `compare_specs`, but it doesn't actually affect the way the comparison is done. All we really need from it is the headers.

```suggestion
        headers = [args.args[0], args.args[1]]
```